### PR TITLE
Remove unsafe code

### PR DESCRIPTION
**Pull request by JasonLG1979**
As mentioned in https://github.com/librespot-org/librespot/discussions/917#discussioncomment-2007452.

### CODE REVIEW

**Comment by JasonLG1979** [comment]
Still need to update the doc strings. I need to read up on the conventions.

**Comment by roderickvd** [comment]
Ready?

File: CHANGELOG.md [hunk 1]
```diff
@@ -28,6 +28,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - [main] Prevent hang when discovery is disabled and there are no credentials or when bad credentials are given.
 - [main] Don't panic when parsing options. Instead list valid values and exit.
 - [main] `--alsa-mixer-device` and `--alsa-mixer-index` now fallback to the card and index specified in `--device`.
+- [spotify_id] Removed unsafe code.
```
Comment (roderickvd) [inline CHANGELOG.md:31]: This should read `[core]` and `(breaking)`.

File: core/src/spotify_id.rs [hunk 2]
```diff
@@ -136,15 +137,15 @@ impl SpotifyId {
 
     /// Returns the `SpotifyId` as a base16 (hex) encoded, `SpotifyId::SIZE_BASE16` (32)
     /// character long `String`.
-    pub fn to_base16(&self) -> String {
+    pub fn to_base16(&self) -> Result<String, FromUtf8Error> {
```
Comment (roderickvd) [inline core/src/spotify_id.rs:140]: I prefer this to return the ubiquitous `librespot_core::Error` type. (Wait, is this PR against `dev` or `new-api`? Let's make it the latter.)

**Review by roderickvd — COMMENTED** [review]
Sorry for the mix-up but let's make sure this targets `new-api` so we can use the new ubiquitous `librespot_core::Error` type.

**Comment by JasonLG1979** [comment]
> Sorry for the mix-up but let's make sure this targets `new-api` so we can use the new ubiquitous `librespot_core::Error` type.

I can do both `dev` and `new-api` so users of `dev` (me and Raspotify) can benefit since we have no ETA on the `new-api` branch.

There's also the logic that states that since `dev` is the main branch and `new-api` is a branch off of that, that it would be up to whoever (me) to manage the conflicts with the `new-api` branch when it is rebased.

My proposal is to merge this into `dev` and then I will rebase `new-api` and fix any conflicts.

File: CHANGELOG.md (hunk 1 above)
Comment (JasonLG1979) [reply CHANGELOG.md:31]: I can do that.

**Comment by JasonLG1979** [comment]
The general idea being that if code can benefit `dev` without major refactoring it should be merged into `dev` 1st and then `new-api` should be rebased.

**Comment by JasonLG1979** [comment]
I'm also having an issue with compiling the `new-api` branch for armv6 (required for Pi 1 and Pi Zero). I can't remember the exact error off the top of my head and I haven't really looked that far into it yet but as it sits you can't use the `make` script in the Raspotify repo to build a armv6 binary like you can the dev branch. That will certainly affect other upstream devs if it turns out that the `new-api` branch simply just isn't compatible with armv6.

**Comment by roderickvd** [comment]
> I can do both `dev` and `new-api` so users of `dev` (me and Raspotify) can benefit since we have no ETA on the `new-api` branch.

I am considering doing an intermediate merge of `new-api` to `dev` soon-ish. So we can get all of the HTTP metadata and audio retrieval out there, before we take the plunge to rip out `spirc` and implement the new `dealer`. Symphonia should be releasing 0.5 soon with proper gapless support and I want to wait for that.

Also I am porting the recent `librespot-java` work for that extra metadata over HTTP.

> There's also the logic that states that since `dev` is the main branch and `new-api` is a branch off of that, that it would be up to whoever (me) to manage the conflicts with the `new-api` branch when it is rebased.
> 
> My proposal is to merge this into `dev` and then I will rebase `new-api` and fix any conflicts.

OK, sounds good to me.

> I'm also having an issue with compiling the `new-api` branch for armv6 (required for Pi 1 and Pi Zero). I can't remember the exact error off the top of my head and I haven't really looked that far into it yet but as it sits you can't use the `make` script in the Raspotify repo to build a armv6 binary like you can the dev branch. That will certainly affect other upstream devs if it turns out that the `new-api` branch simply just isn't compatible with armv6.

OK, let me know. Assuming that's the case (I don't have any of such devices) I don't know what I did to cause that.

**Comment by JasonLG1979** [comment]
> I am considering doing an intermediate merge of new-api to dev soon-ish. So we can get all of the HTTP metadata and audio retrieval out there, before we take the plunge to rip out spirc and implement the new dealer. Symphonia should be releasing 0.5 soon with proper gapless support and I want to wait for that.

> Also I am porting the recent librespot-java work for that extra metadata over HTTP.

Sounds good.

> OK, let me know. Assuming that's the case (I don't have any of such devices) I don't know what I did to cause that.

It's nothing you did directly one of the dependencies won't compile. It's nothing in `librespot`'s actual code base. 

Like I said I haven't really looked that hard into it and I haven't tried to compile that branch on an actual Pi Zero because it takes like 4 hours. It could very well be just a matter of rejiggering some tooling.

**Comment by roderickvd** [comment]
Merged. As a suggestion for `new-api`, you may want to again remove the `unsafe` blocks and then do the fix-until-it-compiles dance, instead of trying to merge this. The player event handler didn't change, but metadata did, and I don't think it would be pretty. But it's just a suggestion.

**Comment by JasonLG1979** [comment]
> Merged. As a suggestion for `new-api`, you may want to again remove the `unsafe` blocks and then do the fix-until-it-compiles dance, instead of trying to merge this. The player event handler didn't change, but metadata did, and I don't think it would be pretty. But it's just a suggestion.

My general mode of operation is basically what you describe.

**Pull request merged**